### PR TITLE
August Improvements

### LINKED ISSUE #184: Editing interstate assignments doesn't work properly
This particularily applies when ADDING new assignments.

### LINKED ISSUE #187: Zoom-to-element margins should not be 0
When zooming to an element, such as when clicking a search result or the outline, the view should not zoom all the way into the element with no margin. A small margin should be kept around the element.

### LINKED ISSUE #188: A close button on the minimap should be available to disable the minimap

### LINKED ISSUE #189: Replace the clear info button in the details panel with a button to hide the panel

### LINKED ISSUE #190: A button in the details panel should allow toggling between horizontal and vertical layouts.

### LINKED ISSUE #192: Group transformations by type
Transformations should be grouped together by types into collapsible groups.
Next to each group of the same type, instead of quick-apply it should have a button to quick-apply ALL transformations in the given group.

### LINKED ISSUE #193: Clicking a transformation should zoom to the affected elements

### LINKED ISSUE #196: Allow configuring a default custom transformation directory for workspaces
When the extension starts up, this folder will always be searched for custom transformations.

### LINKED ISSUE #197: Fix / improve goto source and generated code
Both features are currently unreliable and often don't find the correct files. This is particularly true for remote settings (WSL, containers, SSH)

### PR DESCRIPTION
**Pull request by phschaad**
- [x] Add a disable minimap button (Closes #188 )
- [x] Group transformations by type and allow applying of multiple transformations of the same type (Closes #192 )
- [x] Zoom to affected elements when selecting a transformation (Closes #193 )
- [ ] ~Integrate the passes pipeline into the transformation list~ Moved to September round.
- [x] Allow toggling between the vertical and horizontal layout through the UI (Closes #190 )
- [x] Allow toggling and folding of the details panel (Closes #189 )
- [x] Fix several issues related to metadata and property editing (Closes #184 )
- [x] Allow configuring of default custom transformation directories (Closes #196 )
- [x] Fix / improve goto source and generated code behavior (Closes #197 )
- [ ] ~Fix issues with the transformation list refreshing~ Moved to September round.

Companion PR https://github.com/spcl/dace-webclient/pull/89 closes #187.

NOTE: Only one improvements patch is planned for the month of August. This is an exception.

### CODE REVIEW

File: src/components/sdfg_viewer.ts [hunk 1]
```diff
@@ -220,17 +242,23 @@ export class SdfgViewerProvider
...
+                // If the message passes a cache path then use that path,
+                // otherwise reconstruct the folder based on the default cache
+                // directory with respect to the opened workspace folder and the
+                // SDFG name.
                 let cachePath: string = message.cachePath;
                 if (!cachePath) {
-                    const SdfgFileName = DaCeVSCode.getInstance()
-                        .getActiveSdfgFileName();
-                    if (!SdfgFileName)
+                    if (vscode.workspace.workspaceFolders) {
```
Comment (tbennun) [inline src/components/sdfg_viewer.ts:251]: what if the SDFG already provides an absolute path to the cpp code?

File: src/utils/sdfg_meta_dict.json [hunk 2]
```diff
@@ -744,6 +746,12 @@
       "category": "General",
       "default": 0
     },
+    "optional": {
```
Comment (tbennun) [inline src/utils/sdfg_meta_dict.json:749]: where is `pool` from dace 0.14?

File: src/webclients/components/sdfv/properties/properties.ts [hunk 3]
```diff
@@ -99,6 +111,14 @@ export class KeyProperty {
     ) {
     }
 
+    public markDeleted(): void {
```
Comment (tbennun) [inline src/webclients/components/sdfv/properties/properties.ts:114]: what's this for?

File: src/webclients/components/sdfv/transformation/transformation.ts [hunk 4]
```diff
@@ -261,80 +261,83 @@ export function showTransformationDetails(xform: any): void {
...
         'text': 'Preview',
     })).appendTo(xformButtonContainer);
 
     $('<div>', {
-        'class': 'button',
-        'click': () => {
-            applyTransformation(xform);
+        class: 'button',
+        click: () => {
+            applyTransformations(xform);
```
Comment (tbennun) [inline src/webclients/components/sdfv/transformation/transformation.ts:335]: xform or [xform]?

Comment (tbennun) [reply src/webclients/components/sdfv/transformation/transformation.ts:335]: ah, it's `...xforms`. disregard.

File: src/webclients/components/sdfv/utils/attributes_table.ts [hunk 5]
```diff
@@ -404,15 +411,19 @@ export function attrTablePutDict(
...
+                class: 'row attr-table-row',
+            }).appendTo(rowbox);
+            const attrProp = await attributeTablePutEntry(
+                k, v, valMeta, val, elem, xform, row, true, false, true
             );
 
             if (attrProp.deleteBtn)
                 attrProp.deleteBtn.on('click', () => {
                     attrProp.keyProp?.getInput().val('');
+                    attrProp.keyProp?.markDeleted();
```
Comment (tbennun) [inline src/webclients/components/sdfv/utils/attributes_table.ts:426]: why/when would I want to erase a property?

File: src/webclients/components/sdfv/utils/helpers.ts [hunk 6]
```diff
@@ -301,6 +302,35 @@ export function elementUpdateLabel(
...
+                            element.data.attributes.condition.string_data;
+                        if (strdata !== '1' && strdata !== 1 &&
+                            strdata !== 'true' && strdata !== true)
+                            condition = strdata;
+                    }
+                    let assignments = null;
+                    if (element.data.attributes?.assignments) {
+                        const assDict = element.data.attributes.assignments;
+                        const assList = [];
+                        for (const k in assDict)
```
Comment (tbennun) [inline src/webclients/components/sdfv/utils/helpers.ts:320]: oh no

**Review by tbennun — CHANGES_REQUESTED** [review]
Minor changes and questions only

File: src/webclients/components/sdfv/properties/properties.ts (hunk 3 above)
Comment (phschaad) [reply src/webclients/components/sdfv/properties/properties.ts:114]: See below

File: src/webclients/components/sdfv/utils/attributes_table.ts (hunk 5 above)
Comment (phschaad) [reply src/webclients/components/sdfv/utils/attributes_table.ts:426]: This is used for dictionaries and data descriptors. If a data descriptor is to be removed, the corresponding property is deleted. The terminology may be slightly misleading (unintentionally, this crept in) - a property here is not equal to an SDFG property. It is more general here, so even elements of a dictionary (or dict property) are called properties.

File: src/webclients/components/sdfv/utils/helpers.ts (hunk 6 above)
Comment (phschaad) [reply src/webclients/components/sdfv/utils/helpers.ts]: Oh dear..

File: src/webclients/components/sdfv/utils/attributes_table.ts (hunk 5 above)
Comment (tbennun) [reply src/webclients/components/sdfv/utils/attributes_table.ts:426]: Needs a brief comment somewhere

**Review by tbennun — APPROVED** [review]

**Pull request merged**